Tracing

systay opened the pull request:
Using these changes, this trace was captured:
![Trace](https://i.imgur.com/wnlYkZt.png)

deepthi commented:
Nice! The actual choice of tracing tool should be hidden from all other packages besides trace so that we don't have refs to Jaeger sprinkled around the code. That will make it easier for any future refactor that changes the tracing tool used.

derekperkins commented:
I love this. My only worry is that OpenTracing is the wrong instrumentation library. IMO, the biggest issue is that OpenTracing doesn't implement a wire spec (https://github.com/opentracing/specification/issues/64), and so various vendor implementations are incompatible. Each user that wants a new exporter will have to add specific support to Vitess, and then each of those would have its own flags for configuration. Here's an example in practice: https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/. That adds significant dependency bloat.

I think we should implement this with OpenCensus as opposed to OpenTracing. It is self-contained and we wouldn't need to add any vendor support. Users are able to wire up their own exporters inside their own application logic without having to change any Vitess configuration. 

OpenCensus is instrumented by default in all the Google Cloud libraries and I believe in all the Azure libraries as well. Specifically for Go, OC also tags pprof traces that would help with performance optimizations.

For @systay specifically and this Jaeger implementation, I'm not sure wiring up OpenCensus inside of Vitess would interop with your existing Jaeger infrastructure that uses OpenTracing. Do you have any suggestions @rakyll?

As a side note, this is more reason to promote gRPC as the preferred protocol over the MySQL wire protocol.

derekperkins commented:
dgraph just integrated OpenCensus here
https://github.com/dgraph-io/dgraph/pull/2739

leoxlin commented on java/jdbc/src/main/java/io/vitess/jdbc/VitessVTGateManager.java:
```diff
@@ -174,45 +178,56 @@ private static void actuallyCloseRefreshedConnection(final VTGateConnection old)
   }
 
+  private static TlsOptions getTlsOptions(VitessConnection con) {
+    String keyStorePath = nullIf(con.getKeyStore(), getProperty(Property.KEYSTORE_FULL));
```
:nit: If we are already changing this. A more java 8 way to do this would be:

```
      Optional
          .ofNullable(con.getKeyStore())
          .orElse(getProperty(Property.KEYSTORE_FULL))
```

systay replied:
Meh... Call me old-school, but I like that my way of doing it is shorter and no extra objects are created.

leoxlin commented on java/jdbc/src/main/java/io/vitess/jdbc/VitessVTGateManager.java:
```diff
@@ -174,45 +178,56 @@ private static void actuallyCloseRefreshedConnection(final VTGateConnection old)
+      TlsOptions tlsOptions = getTlsOptions(connection);
+      RpcClient rpcClient = new GrpcClientFactory(retryingConfig)
+          .createTls(context, hostInfo.toString(), tlsOptions);
+      String path = tlsOptions.getTrustStore().getPath();
```
The path variable is not needed. Just create this in the construction invocation like the other path `tlsOptions.getKeyStore().getPath()`

leoxlin commented on java/grpc-client/src/main/java/io/vitess/client/grpc/GrpcClientFactory.java:
```diff
@@ -88,9 +90,11 @@ public GrpcClientFactory setNameResolverFactory(NameResolver.Factory value) {
    */
   @Override
   public RpcClient create(Context ctx, String target) {
+    ClientTracingInterceptor tracingInterceptor = new ClientTracingInterceptor();
```
Can we add an enable flag for this and default to false. `public GrpcClientFactory enableOpenTracing()`
This way we can toggle it with a jdbc config as well.

systay replied:
I'm not sure what you mean here. Would you mind coding up an example of what you are after?

inexplicable commented:
> I love this. My only worry is that OpenTracing is the wrong instrumentation library. IMO, the biggest issue is that OpenTracing doesn't implement a wire spec ([opentracing/specification#64](https://github.com/opentracing/specification/issues/64)), and so various vendor implementations are incompatible. Each user that wants a new exporter will have to add specific support to Vitess, and then each of those would have its own flags for configuration. Here's an example in practice: https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/. That adds significant dependency bloat.
> 
> I think we should implement this with OpenCensus as opposed to OpenTracing. It is self-contained and we wouldn't need to add any vendor support. Users are able to wire up their own exporters inside their own application logic without having to change any Vitess configuration.
> 
> OpenCensus is instrumented by default in all the Google Cloud libraries and I believe in all the Azure libraries as well. Specifically for Go, OC also tags pprof traces that would help with performance optimizations.
> 
> For @systay specifically and this Jaeger implementation, I'm not sure wiring up OpenCensus inside of Vitess would interop with your existing Jaeger infrastructure that uses OpenTracing. Do you have any suggestions @rakyll?
> 
> As a side note, this is more reason to promote gRPC as the preferred protocol over the MySQL wire protocol.

thanks a bunch for doing this! just wanna echo ☝️, as we're looking for a better tracing from pinterest's vitess client as well, and our current tracing setup is strongly biased towards `zipkin`, so if the Jaeger impl could be wrapped as a `SpanFactory` and init via `RegisterSpanFactory`, it might facilitate a lot more adoptions.

one more note, as we're using the mysql protocol as of now, we were wondering if `plugin_mysql_server#ComQuery` could use sql comments (maybe new directive) to provide the trace context mimicing what grpc api provides (my guess there're still lots of mysql protocol users of vitess)

systay commented:
@inexplicable I've reworked the PR to make it easy to add more backends. Basically you'd do something similar to [this]( https://github.com/vitessio/vitess/pull/4610/commits/0692ca81917368b80d43334084aeec050a552fec), and then hook it up like [this](https://github.com/vitessio/vitess/pull/4610/commits/fbd45d179045a6ea7a30e856262b53a6d49076cd#diff-7b20f10d47098dcf86bc7c996e92db78R133)

WDYT?

systay commented:
@inexplicable wrote:
> one more note, as we're using the mysql protocol as of now, we were wondering if plugin_mysql_server#ComQuery could use sql comments (maybe new directive) to provide the trace context mimicing what grpc api provides (my guess there're still lots of mysql protocol users of vitess)

This sounds like a really great idea! Once this gets merged, we'll see if anyone is interested in implementing your suggestion.

inexplicable commented on go/trace/trace.go:
```diff
@@ -97,17 +101,31 @@ func RegisterSpanFactory(sf SpanFactory) {
-func (fakeSpan) Annotate(string, interface{}) {}
+// StartTracing enables tracing for a named service
+func StartTracing(serviceName string) io.Closer {
+  switch *tracingServer {
```
alternatively, we could use `package init` to register from each tracingService, and avoid the need to explicitly change the following code. 

```
var tracingServices map[string]func()io.Closer

// in fake.go
func init(){
 tracingServices['fake'] = func(){return &nilCloser{}}
}

// in jaeger.go
func init(){
 tracingServices['jaeger']=func(){
    tracer, closer, err := NewJagerTracerFromEnv(serviceName)
    if err != nil {
      log.Error(vterrors.Wrapf(err, "failed to create a jaeger tracer"))
      return &nilCloser{}
    }
    // Register it for all openTracing enabled plugins, mainly the grpc connections
    opentracing.SetGlobalTracer(tracer)

     // Register it for the internal Vitess tracing system
    RegisterSpanFactory(OpenTracingFactory{Tracer: tracer})
    return closer
 }
}
```

so that for a new tracing service, like `zipkin`, i could add just a `zipkin.go` that has a similar `init`

inexplicable replied:
this is exciting! thank u so much for taking this closer to life! i was wondering if zipkin could be seamlessly integrated with `opentracing api`, but then i've found: https://github.com/openzipkin-contrib/zipkin-go-opentracing

so i guess i'm happy with the choice at least, wish i could speak for the more general audience 🤣

systay replied:
Neat. Yeah, your `init()` solution looks way better. I took it and changed it a bit but it's the same idea. Thanks!

dweitzman commented:
I don't have any pre-existing context on opentracing or opencensus, but poking a little bit at opencensus does seem more forward-thinking than opentracing.

For example, I notice that opencensus has support for establishing a link from a metric bucket to a sample trace from that bucket (a.k.a. an exemplar). Exemplars are super duper useful for investigating things like tail latency in monitoring graphs. Dunno if what open source tooling supports them, but this video shows an example of how exemplars are used to investigate incidents at Google: https://youtu.be/U72b4Nl0Ftw?t=1290

deepthi commented on go/trace/trace.go:
```diff
@@ -81,12 +89,14 @@ func CopySpan(parentCtx, spanCtx context.Context) context.Context {
-	New(parent Span) Span
-	FromContext(ctx context.Context) (Span, bool)
-	NewContext(parent context.Context, span Span) context.Context
+  New(parent Span, label string, spanType SpanType) Span
```
Need comments describing interface methods

systay commented on go/cmd/vtclient/vtclient.go:
```diff
@@ -141,6 +142,9 @@ func run() (*results, error) {
 	args := flag.Args()
 
+	closer := trace.StartTracing("vtclient")
+	defer trace.LogErrorsWhenClosing(closer)
```
not sure what you mean here. Are you thinking about `span.Finish()`? `trace.StartTracing("vtclient")` starts tracing on the currently running process, and identifies the process by name. The closer makes sure spans are flushed before the process finishes. It's preferable to close the tracing like this, but not critical if it's not done.

Spans otoh, only exist for the duration of an operation, and must be opened and closed in pairs.

deepthi replied:
it makes sense now.

systay commented on go/cmd/vtclient/vtclient.go:
```diff
@@ -141,6 +142,9 @@ func run() (*results, error) {
 	flag.Parse()
 	args := flag.Args()
 
+	closer := trace.StartTracing("vtclient")
```
This is not starting a span, it's starting tracing on the current process.

dweitzman replied:
Starting tracing seems like something `go/cmd/vtclient/plugin_opentracing.go` should do, for consistency with how we handle metrics (`go/cmd/vtctld/plugin_prometheusbackend.go` would be an example)

The tricky thing here seems like the closing, since there's not yet any `go/vt/clientenv` yet as a parallel to `go/vt/servenv` which could let a plugin register shutdown hooks across all the client binaries.

It looks like most vitess client binaries do `defer logutil.Flush()` and `defer exit.RecoverAll()` or `exit.Recover()` (from `go/exit`) at the beginning.

Probably the "right" thing to do is to create a more formal `go/vt/clientenv` with pluggable exit hooks, but for the short term maybe just having `defer trace.Flush()` or `defer trace.Close()` or something like could be fine.

systay commented on go/trace/jaeger.go:
```diff
@@ -0,0 +1,98 @@
+// JAEGER_PASSWORD
+// JAEGER_AGENT_HOST
+// JAEGER_AGENT_PORT
+func NewJagerTracerFromEnv(serviceName string) (opentracing.Tracer, io.Closer, error) {
```
👍

systay commented on go/trace/trace.go:
```diff
@@ -20,34 +20,51 @@ limitations under the License.
+
+// NewClientSpan returns a span and a context to register calls to dependent services
+func NewClientSpan(inCtx context.Context, serviceName, spanLabel string) (Span, context.Context) {
+  span, ctx := NewSpan(inCtx, spanLabel, Client)
```
This is in place so that services we are calling out to that do not themselves support tracing are still traced as such. "peer.service" is not a jaeger thing, it's an OpenTracing thing: https://github.com/opentracing/specification/blob/master/semantic_conventions.md

systay commented on go/trace/trace.go:
```diff
@@ -97,17 +107,43 @@ func RegisterSpanFactory(sf SpanFactory) {
-func (fakeSpan) Finish()                      {}
-func (fakeSpan) Annotate(string, interface{}) {}
+// StartTracing enables tracing for a named service
+func StartTracing(serviceName string) io.Closer {
```
I think I need help with the commenting of this method. 

This is not about starting a span, it's about starting up the tracing system. It's done like this so that executables can tell the tracing system which name they are under ("vtgate", "vtclient", "vtwork", etc), and a chance of explicitly closing the tracing system so that everything is flushed nicely.

systay commented on go/trace/utils.go:
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2017 Google Inc.
```
I don't mind either way. This is what the existing files in the trace package had, so I just kept it. Maybe it's time to check this automatically so that all files look the same?

systay commented on go/vt/schemamanager/schemaswap/schema_swap.go:
```diff
@@ -308,7 +308,7 @@ func (schemaSwap *Swap) executeSwap() error {
 	errHealthWatchers := schemaSwap.runOnAllShards(
 		func(shard *shardSchemaSwap) error {
-			return shard.startHealthWatchers()
+			return shard.startHealthWatchers(schemaSwap.ctx )
```
good catch!

systay commented on go/vt/grpcclient/client.go:
```diff
@@ -98,6 +100,14 @@ func Dial(target string, failFast FailFast, opts ...grpc.DialOption) (*grpc.Clie
 		newopts = append(newopts, grpc.WithStreamInterceptor(grpc_prometheus.StreamClientInterceptor))
 	}
 
+	// Enable trace information, if we have a real tracer
```
👍

systay commented:
Hi @deepthi! 👋 

I think I have addressed all your concerns. If I've missed anything, please let me know what, and I'll take a look straight away

deepthi commented:
> Hi @deepthi! 
> 
> I think I have addressed all your concerns. If I've missed anything, please let me know what, and I'll take a look straight away

LGTM. We'll wait for @dweitzman to look over the new changes, and also to get consensus on the slack conversation.

dweitzman commented on go/vt/vtgate/executor.go:
```diff
@@ -122,6 +123,11 @@ func NewExecutor(ctx context.Context, serv srvtopo.Server, cell, statsName strin
 
 // Execute executes a non-streaming query.
 func (e *Executor) Execute(ctx context.Context, method string, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable) (result *sqltypes.Result, err error) {
+	span, ctx := trace.NewClientSpan(ctx, "mysql", "executor.Execute")
```
My inclination is that `NewClientSpan` doesn't need to exist at this time and omitting the `peer.service` tag seems like it probably shouldn't cause any harm?

Different tracing services have different ways of describing remote endpoints (for example, zipkin: https://godoc.org/github.com/openzipkin/zipkin-go/model#Endpoint).

Rather than trying to figure out how to accommodate them now, seems like we can leave them untagged. The "mysql" here adds no additional information, since it's implied by the span name.

systay replied:
I rather leave it in for now. This is the way OpenTracing annotates peer services [1](https://github.com/opentracing/specification/blob/master/semantic_conventions.md) and not a Jaeger specific thing. In our tracing, marking service such as MySQL and Zookeeper as such is helpful, instead of having to infer it from the method that is running.

sougou commented:
Chiming in here to resolve some discussions (including some from slack):
1. Now that OpenCensus and OpenTracing have agreed to consolidate, we're good to continue with the current `OpenTracing` implementation. We owe ourselves a promissory note to change to the consolidated API once it's finalized.
2. Ideally, the code should not depend on Jaeger. The way to achieve this is to have all the code depend on `OpenTracing`, and then we add a 'plugin' file, like here: https://github.com/vitessio/vitess/blob/master/go/cmd/vtgate/plugin_prometheusbackend.go, that will bring in Jaeger.
3. We can submit the plugin file only if jaeger does not cause binary bloat. If there is bloat, then there should be no plugin file. Those who want jaeger (or any other tracing) will need to pull the source, add the necessary plugin and build.

I'll let @deepthi and @dweitzman do review the rest of the code.

systay commented:
:wave: Hi all!

Thanks @sougou.

To summarize: 
* OpenTracing vs OpenSensus - Not an issue anymore. We continue with this PR and change if/when we know we need to. Correct @derekperkins?

* Move jaeger to plugin files - I believe this has been done - not sure if `vendor.json` needs to be handled some special way when we are dealing with plugins. @deepthi - am I doing it right?

* @dweitzman I didn't see an answer to your `trace.NewClientSpan` review. Is having it in there a blocker for your acceptance or is this something we can have for now and maybe clean out when moving to the new joint project?

* Around binary bloat, this is what my local builds say:

_without tracing:_
```
vitess/base         4.59GB
vitess/bootstrap    2.58GB
```

_with tracing:_
```
vitess/base         4.66GB
vitess/bootstrap    2.58GB
```

I get that to 0.07 GB with jaeger and opentracing linked in. Doesn't sound too bad to me, but I don't know what's acceptable.

derekperkins commented:
At a high level, yes, let's merge this somewhat provisionally, knowing that there may be breaking changes when the OpenCensus/OpenTracing merger matures.

As for bloat, I don't think anyone is running base or bootstrap in production. I'd be more interested in the individual binary changes. If 70 MB is the increase including source and every binary, meaning a few MB per binary, then I don't think it's a problem.

dweitzman commented on go/trace/trace.go:
```diff
@@ -79,35 +88,74 @@ func CopySpan(parentCtx, spanCtx context.Context) context.Context {
-// fakeSpan implements Span with no-op methods.
-type fakeSpan struct{}
+var (
+	tracingServer = flag.String("tracer", "noop", "tracing service to use.")
```
Nit: the punctuation as the end of the flag description seems unnecessary, although there are some vitess flags that have that

dweitzman commented on go/trace/trace.go:
```diff
@@ -79,35 +88,74 @@ func CopySpan(parentCtx, spanCtx context.Context) context.Context {
+		}
+
+		altStr := strings.Join(options, ", ")
+		log.Error(vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "no such tracing service found. alternatives are: %v", altStr))
```
To my taste, the use of `vterrors.Errorf` here feels unnecessary. An alternative might be something like:

`log.Errorf("skipping initialization of unknown tracing service: %v. Valid options: %v", *tracingServer,  strings.Join(options, ", "))`

dweitzman commented on go/trace/trace_test.go:
```diff
@@ -17,22 +17,134 @@ limitations under the License.
 }
+
+func (f *fakeTracer) assertNoSpanWith(t *testing.T, substr string) {
+	for _, logLine := range f.log {
```
Probably good to call `t.Helper()` at the top of this function so any errors report the caller as the failed test instead of naming the helper function.

dweitzman commented on go/vt/worker/instance.go:
```diff
@@ -129,6 +130,9 @@ func (wi *Instance) setAndStartWorker(ctx context.Context, wrk Worker, wr *wrang
 	go func() {
 		log.Infof("Starting worker...")
+		span, ctx := trace.NewSpan(wi.currentContext, "work")
+		span.Annotate("extra", wrk.StatusAsText())
```
vtworker status can be quite big. Might be excessively large to be an annotation, and its value is also somewhat opaque and non-specific which doesn't feel very annotation-like.

dweitzman commented on vendor/vendor.json:
```diff
@@ -900,6 +936,114 @@
+		{
+			"checksumSHA1": "H81V2Jfwte/WYslbaG9ykrD1+Q8=",
+			"path": "github.com/uber/jaeger-client-go",
+			"revision": "3231c1a9c59534e968cd021da5e283cbd0171397",
```
Maybe use a tagged released version like v2.16.0 == 2f47546e3facd43297739439600bcf43f44cce5d here?

Go modules have a preference for released versions, so preferring released versions where possible in vendor.json may ease that transition when the time comes.

Same comment for other deps here that have recent enough releases.

dweitzman commented on go/cmd/vtclient/plugin_opentracing.go:
```diff
@@ -0,0 +1,28 @@
+/*
+Copyright 2018 Google Inc.
```
I think "Copyright 2019 The Vitess Authors." is the new copyright header.

dweitzman commented:
The `useTracing` JDBC connection parameter feels somehow off, maybe because it's inflexible by assuming opentracing is the only tracing implementation.

I'm not sure the impact of defaulting to true. Dunno how jars are packaged and dependencies fit into all this, since my folks don't (yet) use the grpc driver.

I would prefer to have someone who knows and cares about the grpc driver review that part. Moving the java stuff to a separate PR and landing the golang stuff now might be a reasonable strategy.

systay commented:
Hi @dweitzman (and others watching)

This whole PR assumes OpenTracing. If you are not using OT, you'll not be able to follow spans from your application in to Vitess. Luckily, OT and OpenCensus have recently announced that they are joining forces, so having to choose between the two should not be an issue going forward.

I added the capability of turning off tracing on the JDBC end altogether by adding `useTracing=off` to the JDBC connection string - but not that this is not really necessary. The default behavior is that tracing the JDBC connection only happens if you already have OT up and running and configured on the host JVM.

dweitzman commented:
I'm happy to defer to the people who have stronger involvement with the Java client stuff, if folks who use and care about the grpc driver are comfortable with this being on-by-default and not configurable to use other systems (opencensus and opentracing intend to merge, but they have not merged).

My own inclinations are:
- something `tracer=opentracing` seems more future-proof than `useTracing=true`, for the same reasons in Java as in Go
- off by default seems like a safer place to start than on by default

sougou commented:
It's likely that we'll have to make changes going forward, but I think what we have here is in a usable state. So, I'd like to get this merged. I'll wait for a little bit longer to see if there are any other objections.

leoxlin commented:
I m with @dweitzman on the naming of the flag.

Maybe `tracer=none` and `tracer=opentracing` to turn it on. Just for future proofing people's configs.

systay commented:
Hmm... Not sure about the naming of parameters. On the go side, you do not get to choose between OpenTracing and something else. You only get to choose which OT back-end to use. Jaeger can be used as a backend in OpenCensus mode, or you can use the OT API to talk with Jaeger.

There is also a difference between go and Java in that in the go side, we are hosting the process, and so need to make a decision on how to configure and start tracing, whereas on Java land, we are making a library loaded by someone else's process that already has tracing configured.

So, if we want to future proof the arguments we use, I think we should use what you suggest in the JDBC connection string, and change the `go` params to include `ot` or `oc` in the name. 

TL;DR;

* _Java_: `tracing=[off,opentracing]`
* _go_: `tracing=[off,opentracing_jaeger]`

We've strayed far from code quality into the highly subjective API design land (here be dragons :) ). Anyone from Planetscale have opinions on this? @sougou @deepthi

dweitzman commented:
> TL;DR;
> 
> * _Java_: `tracing=[off,opentracing]`
> * _go_: `tracing=[off,opentracing_jaeger]`
> 
> We've strayed far from code quality into the highly subjective API design land (here be dragons :). Anyone from Planetscale have opinions on this? @sougou @deepthi

sgtm

systay commented:
Ok. I think this PR is ready to be merged now. I think I have addressed all concerns:

* default is no tracing anywhere
* to remove Jaeger binaries from the build, just remove the `go/trace/plugin_jaeger.go` file
* future-proofed command line arguments

I have also tried to add another backend to make sure the infrastructure works. It's just a matter of adding a single file next to the `plugin_jaeger.go` and updating `govendor`. This PR only includes support for Jaeger though.

deepthi commented:
If there are no other comments, I'll go ahead and merge this later today.
